### PR TITLE
create a rails generator to install accessly

### PR DESCRIPTION
**Pull request by ehourigan**
Resolves [[ch9193]](https://app.clubhouse.io/lessonly/story/9193/create-a-rails-generator)

Adds an install generator and test

### CODE REVIEW

File: accessly.gemspec [hunk 1]
```diff
@@ -21,11 +21,12 @@ Gem::Specification.new do |spec|
...
 
-  spec.add_dependency "activerecord", ">= 5.0"
+  spec.add_dependency "activerecord", "~> 5.0"
 
   spec.add_development_dependency "bundler", "~> 1.16"
   spec.add_development_dependency "rake", "~> 10.0"
   spec.add_development_dependency "minitest", "~> 5.0"
   spec.add_development_dependency "database_cleaner"
   spec.add_development_dependency "sqlite3"
+  spec.add_development_dependency "rails", "~> 5.0"
```
Comment (ehourigan) [inline accessly.gemspec]: there's a bug that was recently fixed in rails that supports minitest 5.11.x and generator tests

File: lib/generators/accessly/install/install_generator.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
+require "rails/generators"
+require "rails/generators/active_record"
+
+module Accessly
+  module Generators
+    class InstallGenerator < Rails::Generators::Base
```
Comment (ehourigan) [inline lib/generators/accessly/install/install_generator.rb]: I consulted Rails docs and the thoughtbot/clearance to define the migration generator

File: test/generators/install_generator_test.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+require "test_helper"
+require_relative File.expand_path("../../../lib/generators/accessly/install/install_generator", __FILE__)
+
+class Accessly::InstallGeneratorTest < ::Rails::Generators::TestCase
```
Comment (ehourigan) [inline test/generators/install_generator_test.rb]: this test will create a temp directory inside the root folder of the project, run the generator, and test that the named migration exists... also tests for something namespaced that we didn't create just for sanity of nil/not nil

File: test/generators/install_generator_test.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    FileUtils.rm_rf(destination_root)
+  end
+
+  test "generates migrations" do
+    migration1 = migration_file_name("db/migrate/create_permitted_actions.rb")
+    migration2 = migration_file_name("db/migrate/create_permitted_action_on_objects")
+    bad_migration = migration_file_name("db/migrate/accessly_not_exist")
+
+    migration1.wont_be_nil
+    migration2.wont_be_nil
```
Comment (ehourigan) [inline test/generators/install_generator_test.rb]: @armilam a few PRs ago I introduced `assert_nil` on the deprecated recommendation to use that over `must_equal nil`

`wont_be_nil` and `must_be_nil` can provide the consistency you were looking for and so I plan to open a different pr to change those assertions

File: accessly.gemspec (hunk 1 above)
Comment (armilam) [reply accessly.gemspec]: Does this mean we have to depend on rails? Not that it's a big deal, but it would be neat to depend only on activerecord in case anybody wants to use this gem that's using activerecord but not rails.

Comment (armilam) [reply accessly.gemspec]: Oooh, I see below that we're using Rails's generator stuff, which means that yes, we need to depend on rails.

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class CreatePermittedActionOnObjects < ActiveRecord::Migration<%= migration_version %>
```
Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: This is valid syntax? Whoa.

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+    create_table :accessly_permitted_action_on_objects, force: true, id: :uuid do |t|
+      t.column :segment_id, :integer, default: -1
+      t.column :action, :integer, null: false
+      t.column :actor_id, :integer, null: false
+      t.column :actor_type, :string, null: false
+      t.column :object_type, :string, null: false
+      t.column :object_id, :integer, null: false
+    end
+
+    add_index(:accessly_permitted_action_on_objects, [:segment_id, :action, :actor_id, :actor_type, :object_type, :object_id], unique: true, name: "acessly_paoo_uniq_table_idx")
```
Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: I'm trying to think through some scenarios that would require different indexes. I'll put some ideas in separate threads so we can talk about them separately (does that make sense or should I have just put them all in one thread?).

Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: ### Get a list of users who have permission to edit a particular lesson

#### Why
We might want to show who has permission to edit the lesson so that somebody can possibly choose to revoke somebody's permission.

#### SQL
```sql
SELECT actor_id, actor_type
FROM accessly_permitted_action_on_objects
WHERE segment_id = $1 AND action = $2 AND object_type = $3 AND object_id = $4
```

#### Index that might help
```rb
[:segment_id, :action, :object_type, :object_id]
```

Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: ### Get list of permissions an actor has on a particular lesson

#### Why
When rendering a page, we might want to conditionally render parts based on the different permissions the user has on the lesson. For example, we might render the "Edit" button only if the user has the `:edit` permission, the "Assign" button only if the user has the `:assign` permission, and the Grading Information only if the user has the `:grading` permission.

#### SQL
```sql
SELECT action
FROM accessly_permitted_action_on_objects
WHERE segment_id = $1 AND actor_type = $2 AND actor_id = $3 AND object_type = $4 AND object_id = $5
```

#### Index that might help
```rb
[:segment_id, :actor_type, :actor_id, :object_type, :object_id]
```

Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: ### Get a list of lessons a user has permissions on and the permissions they have on each

#### Why
On the `/contents` page, we'll want to show all the lessons the user has permission to see and perhaps show some additional controls based on their permissions (like an archive button, edit button, etc).

#### SQL
```sql
SELECT object_id, object_type, action
FROM accessly_permitted_action_on_objects
WHERE segment_id = $1 AND actor_type = $2 AND actor_id = $3 AND object_type IN ('Lesson', 'LearningPaths::Path')
```

#### Index that might help
```rb
[:segment_id, :actor_type, :actor_id, :object_type]
```

Actually, the index I proposed previously would also work here:

```rb
[:segment_id, :actor_type, :actor_id, :object_type, :object_id]
```

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_actions.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  def change
+    create_table :accessly_permitted_actions, force: true, id: :uuid do |t|
+      t.column :segment_id, :integer, default: -1
+      t.column :action, :integer, null: false
+      t.column :actor_id, :integer, null: false
+      t.column :actor_type, :string, null: false
+      t.column :object_type, :string, null: false
+    end
+
+    add_index(:accessly_permitted_actions, [:segment_id, :action, :actor_id, :actor_type, :object_type], unique: true, name: "acessly_pa_uniq_table_idx")
```
Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_actions.rb]: Likewise, I'm sure we can come up with more indexes that would be helpful here.

**Comment by ehourigan** [comment]
@armilam should I go ahead and add the proposed indexes to the generator?

**Comment by armilam** [comment]
@ehourigan 

> should I go ahead and add the proposed indexes to the generator?

Only if you agree! I mean to start a discussion with my proposals, but if you're in agreement, then I suppose we might as well add them.

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb (hunk 6 above)
Comment (armilam) [reply lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: @ehourigan I think we should include this index. I believe it will be a common case.

Comment (armilam) [reply lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: @ehourigan I think we should include this index. I believe it will be a common case.

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_actions.rb (hunk 7 above)
Comment (armilam) [inline lib/generators/accessly/install/templates/db/migrate/create_permitted_actions.rb]: @ehourigan I would argue that this index should be:

```rb
[:segment_id, :actor_id, :actor_type, :object_type, :action]
```

Because it seems like we're more likely to get a list of all the actions a user can do or to get a list of all the actions a user can do in a given namespace (object_type) than to get a list of all the actors who can do a particular action.

File: lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb (hunk 6 above)
Comment (ehourigan) [reply lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: we will move action to the end of initial index to cover this usecase

Comment (ehourigan) [reply lib/generators/accessly/install/templates/db/migrate/create_permitted_action_on_objects.rb]: we will add this one with :action at the end
